hail/src/main/scala/is/hail/expr/ir/EmitStream.scala

```diff
@@ -801,7 +897,157 @@ object EmitStream {
+  def kWayZipJoin(
+    mb: EmitMethodBuilder[_],
+    region: Value[Region],
+    streams: IndexedSeq[Stream[PCode]],
+    eltType: PStruct,
+    resultType: PArray,
+    key: IndexedSeq[String]
+  ): Stream[PCode] = new Stream[PCode] {
+    def apply(eos: Code[Ctrl], push: PCode => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[PCode] = {
+      // The algorithm maintains a tournament tree of comparisons between the
+      // current values of the k streams. The tournament tree is a complete
+      // binary tree with k leaves. The leaves of the tree are the streams,
+      // and each internal node represents the "contest" between the "winners"
+      // of the two subtrees, where the winner is the stream with the smaller
+      // current key. Each internal node stores the index of the stream which
+      // *lost* that contest.
+      // Each time we remove the overall winner, and replace that stream's
+      // leaf with its next value, we only need to rerun the contests on the
+      // path from that leaf to the root, comparing the new value with what
+      // previously lost that contest to the previous overall winner.
+
+      val k = streams.length
+      // The leaf nodes of the tournament tree, each of which holds a pointer
+      // to the current value of that stream.
+      val heads = mb.newLocal[Array[Long]]("merge_heads")
+      // The internal nodes of the tournament tree, laid out in breadth-first
+      // order, each of which holds the index of the stream which lost that
+      // contest.
+      val bracket = mb.newLocal[Array[Int]]("merge_bracket")
+      // When updating the tournament tree, holds the winner of the subtree
+      // containing the updated leaf. Otherwise, holds the overall winner, i.e.
+      // the current least element.
+      val winner = mb.newLocal[Int]("merge_winner")
+      val result = mb.newLocal[Array[Long]]("merge_result")
+      val i = mb.newLocal[Int]("merge_i")
+
+      val keyType = eltType.selectFields(key)
+      val curKey = ctx.mb.newPLocal("st_grpby_curkey", keyType)
+
+      val keyViewType = PSubsetStruct(eltType, key: _*)
+      val lt: (Code[Long], Code[Long]) => Code[Boolean] = keyViewType
+        .codeOrdering(mb, keyViewType, missingFieldsEqual = false)
+        .asInstanceOf[CodeOrdering { type T = Long }]
+        .lteqNonnull
+      val hasKey: (Code[Long], Code[Long]) => Code[Boolean] = keyViewType
+        .codeOrdering(mb, keyType, missingFieldsEqual = false)
+        .asInstanceOf[CodeOrdering { type T = Long }]
+        .equivNonnull
+
+      val srvb = new StagedRegionValueBuilder(mb, resultType, region)
+
+      val runMatch = CodeLabel()
+      val LpullChild = CodeLabel()
+      val LloopEnd = CodeLabel()
+      val LaddToResult = CodeLabel()
+      val LstartNewKey = CodeLabel()
+      val Leos = CodeLabel()
+      val Lpush = CodeLabel()
+
+      Code(Leos, eos)
```

Comment:
I don't understand what side effects this statement has that makes it work. It appears that we create a `Code` and then just throw it away. How does `eos` get emitted?

Reply:
Yeah, this syntax is a bit hacky. As part of `CodeBuilder`ifying streams, I have plans to clean this up.

A label is really just an lir block. It starts empty (`Leos._start.first == null`). This line defines the contents of the block (by calling `Leos.end.append(lir.goto(eos.start))`, where `Leos.end == Leos._start` since it's empty), so it is no longer empty. But if `Leos` were never mentioned again, there would be no references to that block after this scope ends, and it would get GCed. But if we jump to this label from some other block, that block will contain a pointer to this one (in the control statement that jumps to the label), so there will be a path from the root block of the method we're building to `Leos._start`.